### PR TITLE
Strategy Design Pattern

### PR DESCRIPTION
**Pull request by katonaaron-msg**
Apply Strategy Design Pattern for Order Creation
- Implement two LocationSelectionAlgorithms and update OrderService to use them.
- Create a bean for the algorithm with LocationSelectionConfiguration
- Read the LocationSelectionAlgorithm implementation's name from the properties file
- Move shippedFrom location from Order to OrderDetail. Migrate DB.

### CODE REVIEW

File: src/main/java/ro/msg/learning/shop/exception/LocationOutOfStockException.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package ro.msg.learning.shop.exception;
+
+import lombok.Getter;
+
+import java.util.UUID;
+
+@Getter
+public class LocationOutOfStockException extends RuntimeException {
+    private final UUID locationId;
```
Comment (cristinaAnca) [inline src/main/java/ro/msg/learning/shop/exception/LocationOutOfStockException.java]: Do you really need these variables? Do you use them elsewhere than the constructor in order to set the error message?

File: src/main/java/ro/msg/learning/shop/model/Location.java [hunk 2]
```diff
@@ -27,4 +29,21 @@ public void setStocks(Set<Stock> stocks) {
         this.stocks = stocks;
         stocks.forEach(stock -> stock.setLocation(this));
     }
+
+    public void removeProductFromStock(@NonNull Product product, @NonNull Integer quantity) {
```
Comment (cristinaAnca) [inline src/main/java/ro/msg/learning/shop/model/Location.java]: Please keep the model entities as POJO (no implementation inside the entities). This method should be part of a service class.

File: src/main/java/ro/msg/learning/shop/service/locationselection/MostAbundantLocationSelectionAlgorithm.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package ro.msg.learning.shop.service.locationselection;
+
+import ro.msg.learning.shop.exception.LocationSelectionException;
+import ro.msg.learning.shop.model.OrderDetail;
+import ro.msg.learning.shop.model.Stock;
+
+import java.util.*;
```
Comment (cristinaAnca) [inline src/main/java/ro/msg/learning/shop/service/locationselection/MostAbundantLocationSelectionAlgorithm.java]: Replace * imports with explicit imports. We don't want to import everything from java.util. Please replace also all the * imports from the entire project with explicit import

File: src/main/resources/application.properties [hunk 4]
```diff
@@ -4,4 +4,6 @@ spring.datasource.password=postgres
 spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.PostgreSQLDialect
 spring.jpa.hibernate.ddl-auto=validate
 spring.jpa.show-sql=true
-spring.jpa.properties.hibernate.format_sql=true
\ No newline at end of file
+spring.jpa.properties.hibernate.format_sql=true
+#ro.msg.learning.shop.location-selection-algorithm=ro.msg.learning.shop.service.locationselection.SingleLocationSelectionAlgorithm
```
Comment (cristinaAnca) [inline src/main/resources/application.properties]: Remove this comment

**Review by cristinaAnca — CHANGES_REQUESTED** [review]

File: src/main/java/ro/msg/learning/shop/exception/LocationOutOfStockException.java (hunk 1 above)
Comment (katonaaron-msg) [reply src/main/java/ro/msg/learning/shop/exception/LocationOutOfStockException.java]: Removed them

**Review by cristinaAnca — APPROVED** [review]

**Pull request merged**